### PR TITLE
Export

### PR DESCRIPTION
**Pull request by eileenmcnaughton**

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@totten OK have at it - but I'm only interested in fixing problems - not improvements :)

File: CRM/Export/BAO/Export.php [hunk 1]
```diff
@@ -760,12 +757,8 @@ public static function exportComponents(
 
     // for CRM-3157 purposes
     $i18n = CRM_Core_I18n::singleton();
-    $outputColumns = array();
-    //@todo - it would be clearer to start defining output columns earlier in this function rather than stick with return properties until this point
-    // as the array is not actually 'returnProperties' after the sql query is formed - making the alterations to it confusing
-    foreach ($returnProperties as $key => $value) {
-      $outputColumns[$key] = $value;
-    }
+    list($outputColumns, $headerRows, $sqlColumns, $metadata) = self::getExportStructureArrays($returnProperties, $query, $phoneTypes, $imProviders, $contactRelationshipTypes, $relationQuery, $selectedPaymentFields);
```
Comment (eileenmcnaughton) [inline CRM/Export/BAO/Export.php]: A disadvantage of this approach is you miss the chance to take advantage of $outputColumns being looped through later & parasiting onto that loop. The advantage however, is that if that loop runs 100,000 times it doesn't have to decide not to extract the headers 999,999 times...

File: CRM/Export/BAO/Export.php [hunk 2]
```diff
@@ -793,125 +786,6 @@ public static function exportComponents(
         //first loop through output columns so that we return what is required, and in same order.
         $relationshipField = 0;
         foreach ($outputColumns as $field => $value) {
-          //we should set header only once
-          if ($setHeader) {
```
Comment (eileenmcnaughton) [inline CRM/Export/BAO/Export.php]: Moved to public static function setHeaderRows which runs before the while loop (only once per column)

File: CRM/Export/BAO/Export.php [hunk 3]
```diff
@@ -1242,11 +1101,6 @@ public static function exportComponents(
         self::mergeSameHousehold($exportTempTable, $headerRows, $sqlColumns, $relationKeyHOH);
       }
 
-      // fix the headers for rows with relationship type
```
Comment (eileenmcnaughton) [inline CRM/Export/BAO/Export.php]: Should we do this once per row .. or perhaps just once per export. Let's go for the latter

File: CRM/Export/BAO/Export.php [hunk 4]
```diff
@@ -2143,4 +2002,196 @@ public static function componentPaymentFields() {
...
+    elseif (substr($field, 0, 5) == 'case_') {
+      if ($query->_fields['case'][$field]['title']) {
+        $headerRows[] = $query->_fields['case'][$field]['title'];
+      }
+      elseif ($query->_fields['activity'][$field]['title']) {
+        $headerRows[] = $query->_fields['activity'][$field]['title'];
+      }
+    }
+    elseif (array_key_exists($field, $contactRelationshipTypes)) {
+      self::manipulateHeaderRows($headerRows, $contactRelationshipTypes);
```
Comment (eileenmcnaughton) [inline CRM/Export/BAO/Export.php]: boring .. once per export

**Comment by eileenmcnaughton** [comment]
I did some profiling on this and I was unable to demonstrate increased performance although i still stand by the principles of this ie moving as much processing out of the dao->Fetch loop as possible - with a view to putting that work into a separate function.

My reading of the profiler output is that the convertToPseudonames function is quite expensive - taking 12-15% of the total time (or 20% of the processing time since only 75% of the time taken in my example is going into the main export work - 2000 rows). 2/3 of that time is being spent on CRM_Core_DAO_Address_construct - which seems wierd - but I was able to know most of that out by adding this

``
class CRM_Core_BAO_Address extends CRM_Core_DAO_Address {

  function __construct() {
    $this->__table = 'civicrm_address';
  }
``

I estimate that on a large export that might be around 12% of the time could be eliminated by blocking the CRM_Core_DAO::__construct function - but in fact the gain may be less as I did my profiling on 4.6 - so I'll try the impact of the set names patch

**Comment by eileenmcnaughton** [comment]
I tried the set names patch on 4.6 and did knock out a 'significant chunk' of the time in convertToPseudonames

I'm unable to fully quantify it because with only 2000 rows the variability between each run is fairly high so I'm focussing on what percent of total time is taken up - but perhaps 6% improvement which is significant but I suspect still more could be done on the convertToPseudoconstant function. However, the overall function needs more clean-up / testing as I believe there is duplicate processing going on between the main loop & the convert function.

**Comment by totten** [comment]
I haven't tested UI, but in terms of being a net improvement to code, +1 for merge. Yaaay unit tests!

**Comment by eileenmcnaughton** [comment]
OK - I tested exporting phone, im, some location fields & there were OK. I think we are good...
